avx2/commitment: F* proofs

W95Psp opened the pull request:
This PR adds proofs by normalization for one avx2 serialization function.

> **(previous text)**
> The main problem here is `libcrux-intrinsics`: the avx2 proofs in ML-KEM relies on very different models.
> I'm not sure what to do here: maybe I should just add a `avx2_extract_minicore.rs` file and add a cfg to choose between either `avx2_extract` or `avx2_extract_minicore`?
> And when we migrate the ML-KEM proofs, we can kill `avx2_extract`.
> 
> Any opinion on that? 

-------

Fixes [#886](https://github.com/cryspen/libcrux/issues/886).
We'll need to file a follow up for the proof of `serialize_6`.

## What does this PR fixes?
 - enables proofs with the Rust minicore
 - bit-level proof of `serialize_4`

Missing:
 - bit-level proof of `serialize_6`
 - higher-level statements / proofs: proofs on integers, not bit vec

## `hax.sh`
Also, in this PR we use `minicore` for proofs.
Our F* models of the Rust `core` library and the extracted code from `minicore` overlap, particularly for `core::arch::*`.
The `minicore` versions offer more accurate and specialized models.

Thus, this PR also adds a utility function (see function `fixup-minicore` in `hax.sh`) that fixes up module names in a given crate: it will make the F* extraction use preferably minicore modules instead of core modules when possible.

This utility function is essentially patching files, which is trivial to do with bash. The former python script was essentially running commands.
Thus I rewrote it in bash.

-------
This following hax PRs were created while working on this libcrux PR:
 - https://github.com/cryspen/hax/pull/1356
 - https://github.com/cryspen/hax/pull/1355
 - https://github.com/cryspen/hax/pull/1354
 - https://github.com/cryspen/hax/pull/1353
 - https://github.com/cryspen/hax/pull/1352

franziskuskiefer commented on libcrux-ml-dsa/src/simd/avx2/encoding/commitment.rs:
```diff
@@ -1,30 +1,69 @@
+            ))]
+    "#)]
+    #[inline(always)]
+    fn aux(simd_unit: Vec256) -> Vec128 {
```
Values should be passed by reference. This style isn't great and may negatively impact performance.
Also, this function needs a proper name that tells the reader what it's doing.

W95Psp replied:
👍

franziskuskiefer commented on libcrux-ml-dsa/hax.py:
```diff
@@ -63,13 +63,13 @@ def __call__(self, parser, args, values, option_string=None) -> None:
             "into",
-            "-i",
-            include_str,
+            # "-i",
+            # include_str,
```
This probably shouldn't be here?

franziskuskiefer commented on libcrux-ml-dsa/hax.py:
```diff
@@ -63,13 +63,13 @@ def __call__(self, parser, args, values, option_string=None) -> None:
             "80",
-            "--interfaces",
-            interface_include,
+            # "--interfaces",
+            # interface_include,
```
This probably shouldn't be here?

franziskuskiefer commented on libcrux-ml-dsa/Cargo.toml:
```diff
@@ -20,7 +20,8 @@ libcrux-sha3 = { version = "0.0.2-beta.3", path = "../libcrux-sha3" }
 libcrux-platform = { version = "0.0.2-beta.3", path = "../sys/platform" }
 libcrux-macros = { version = "0.0.2-beta.3", path = "../macros" }
-hax-lib = { version = "0.1.0", git = "https://github.com/hacspec/hax/"}
+hax-lib = { version = "0.1.0", git = "https://github.com/hacspec/hax/", branch = "replace-body-temp"}
```
The branch needs to go before merging.

franziskuskiefer commented on libcrux-ml-dsa/Cargo.toml:
```diff
@@ -20,7 +20,8 @@ libcrux-sha3 = { version = "0.0.2-beta.3", path = "../libcrux-sha3" }
 libcrux-macros = { version = "0.0.2-beta.3", path = "../macros" }
-hax-lib = { version = "0.1.0", git = "https://github.com/hacspec/hax/"}
+hax-lib = { version = "0.1.0", git = "https://github.com/hacspec/hax/", branch = "replace-body-temp"}
+minicore = { path = "../fstar-helpers/minicore" }
```
So this requires releasing the minicore and actually making it releasable.

W95Psp replied:
Ideally, I'd like to kill that Rust dependency. I have a plan for that but I need to experiment a little: that will require some more rewriting in F*.

W95Psp replied:
So: this dependency is not needed in libcrux-intrsinsics, but I still need those in ml-dsa, but only on specs

franziskuskiefer replied:
Well, it's a dependency, so it needs to be released. `path` doesn't work for that.

franziskuskiefer reviewed (commented):
no review, just a few quick high level comments

W95Psp commented:
I have a regression in my proof somewhere, I need to fix that.
But otherwise this PR is ready for review!

franziskuskiefer commented on libcrux-ml-dsa/hax.sh:
Why does this need to be rewritten and what does it now?

W95Psp replied:
Quoting the commit that introduces this:

> @W95Psp
> [hax: intro hax.sh](https://github.com/cryspen/libcrux/pull/874/commits/c3c24c3322dc21556b9365d2837f7f4aeb30a13a)
> [c3c24c3](https://github.com/cryspen/libcrux/pull/874/commits/c3c24c3322dc21556b9365d2837f7f4aeb30a13a)
> 
> This commit introduces a `hax.sh` script, that aims at replacing the python script.
> 
> The script does a little more: it is fixing up core names into minicore names.
> 
> The union of modules between our hax `core` model (proof-libs in hax) and minicore is not empty. This is a big issue for F*.
> The script thus replace references to every core module provided by minicore into a reference to the corresponding minicore module.

W95Psp replied:
The script is fundamentally orchestrating shell commands (cargo, make, perl, find, etc.). I believe the added logic for patching files and managing file movements is much more natural and concise in Bash. (such operations are native to bash)

Rewriting avoids the overhead of subprocess handling in Python, and leads I believe to a simpler, more maintainable script that better matches its purpose.

franziskuskiefer replied:
That's not a good reason to rewriting entirely good code. I'll accept this for this PR but will not in future.

franziskuskiefer commented on fstar-helpers/minicore/src/abstractions/funarr.rs:
```diff
@@ -0,0 +1,134 @@
+    }
+}
+
+// impl<const N: u64, T> AsRef<[T]> for FunArray<N, T> {
```
Commented code

franziskuskiefer commented on fstar-helpers/minicore/src/abstractions/funarr.rs:
```diff
@@ -0,0 +1,134 @@
+"#
+)]
+#[derive(Copy, Clone, Eq, PartialEq)]
+// pub struct FunArray<const N: usize, T>([T; N]);
```
Commented code

franziskuskiefer commented on fstar-helpers/minicore/src/abstractions/funarr.rs:
```diff
@@ -0,0 +1,134 @@
+#[hax_lib::fstar::replace(
```
What is this module doing? Please add some comments.

franziskuskiefer commented on fstar-helpers/minicore/src/abstractions/bitvec.rs:
```diff
@@ -34,20 +37,81 @@ fn bit_slice_to_string(bits: &[Bit]) -> String {
     }
 }
 
+// impl<const N: u64> FunArray<N, Bit> {
```
Commented code

franziskuskiefer commented on libcrux-ml-dsa/src/simd/avx2/encoding/commitment.rs:
```diff
@@ -1,30 +1,50 @@
+    )
+})]
 #[inline(always)]
+fn serialize_4(simd_unit: &Vec256) -> Vec128 {
```
Did you check that this doesn't impact performance?

W95Psp replied:
No. I am taking a look now then

W95Psp replied:
Ran `cargo criterion` before and after applying my diff to that one file `commitment.rs`.
I got the following very surprising results:
<img width="581" alt="Capture d’écran 2025-04-01 à 15 00 00" src="https://github.com/user-attachments/assets/0bbb2950-f0eb-4005-a182-a1b7d001ae63" />

Seems like the benchs are a bit flaky (as disucssed on gchat with Jonas).
Shall I inverstigate more? (if yes, how?)

W95Psp replied:
Running `cargo bench --bench manual65`:

(seems fine, right?)

# Without commitment.rs patch

(libcrux) KeyGen 65 portable:   132 μs
(libcrux) Sign 65 portable:     395 μs
(libcrux) Verify 65 portable:   123 μs

(libcrux) KeyGen 65 simd256:    130 μs
(libcrux) Sign 65 simd256:      145 μs
(libcrux) Verify 65 simd256:    53 μs

(pqclean) KeyGen 65:    41 μs
(pqclean) Sign 65:      116 μs
(pqclean) Verify 65:    45 μs

# With commitment.rs patch

(libcrux) KeyGen 65 portable:   133 μs
(libcrux) Sign 65 portable:     402 μs
(libcrux) Verify 65 portable:   124 μs

(libcrux) KeyGen 65 simd256:    130 μs
(libcrux) Sign 65 simd256:      144 μs
(libcrux) Verify 65 simd256:    54 μs

(pqclean) KeyGen 65:    42 μs
(pqclean) Sign 65:      116 μs
(pqclean) Verify 65:    46 μs

Second run:

(libcrux) KeyGen 65 portable:   136 μs
(libcrux) Sign 65 portable:     394 μs
(libcrux) Verify 65 portable:   123 μs

(libcrux) KeyGen 65 simd256:    130 μs
(libcrux) Sign 65 simd256:      145 μs
(libcrux) Verify 65 simd256:    53 μs

(pqclean) KeyGen 65:    41 μs
(pqclean) Sign 65:      114 μs
(pqclean) Verify 65:    45 μs

# On main branch

(libcrux) KeyGen 65 portable:   133 μs
(libcrux) Sign 65 portable:     421 μs
(libcrux) Verify 65 portable:   122 μs

(libcrux) KeyGen 65 simd256:    128 μs
(libcrux) Sign 65 simd256:      143 μs
(libcrux) Verify 65 simd256:    53 μs

(pqclean) KeyGen 65:    41 μs
(pqclean) Sign 65:      115 μs
(pqclean) Verify 65:    45 μs

W95Psp commented on libcrux-ml-dsa/src/simd/avx2/encoding/commitment.rs:
```diff
@@ -1,30 +1,50 @@
 use libcrux_intrinsics::avx2::*;
 
+use minicore::abstractions::{bit::Bit, bitvec::BitVec};
```
The dependency exists when extracting to hax.
Shall I add cfg flags then?

karthikbhargavan replied:
Yes, I think to avoid Rust tools from flagging these, let's add `#[cfg(hax)]` guards around such imports, ideally with a comment explaining where they are used.

karthikbhargavan commented:
I will look at this today. Let's try to get it in.

karthikbhargavan reviewed (approved):
I checked the specs and proof and it looks fine.
For future reference, please add a comment above the tactic invocation
to explain what it is doing.

Let's address the other review comments, and get this in.

W95Psp commented on fstar-helpers/minicore/src/core_arch/x86.rs:
```diff
@@ -0,0 +1,635 @@
+    use super::*;
+    #[hax_lib::opaque]
+    pub fn _mm_set_epi8(
+        e15: i8,
```
I fixed the unused args in the two functions.

W95Psp commented:
Thanks for you comments Karthik and Franziskus, I addressed the comments and added a comment about the F* tactic.
I should be good now (modulo bash).

karthikbhargavan commented:
This looks good. I would like to press merge when CI is happy, so that we can get other PRs in.

franziskuskiefer commented:
> @franziskuskiefer shall we merge?

I left another comment. This PR is pretty messy and impossible to review (5! force pushes). That's why I can't really leave an r+ here.
